bin/i18n-map-extractor.js

```diff
@@ -0,0 +1,145 @@
+const recast = require( 'recast' );
+const { forEach, reduce, includes, isEmpty, isFunction } = require( 'lodash' );
+const { readFileSync, writeFileSync } = require( 'fs' );
+const NormalModule = require( 'webpack/lib/NormalModule' );
+
+class wpi18nExtractor {
+	constructor( options ) {
+		const DEFAULT_FUNCTIONS = [
+			'__',
+			'_n',
+			'_x',
+			'_nx',
+			'sprintf',
+		];
+		this.options = options || {};
+		this.options.aliases = this.options.aliases || {};
+		this.options.filename = this.options.filename || 'translation-map.json';
+		this.translationMap = {};
+		this.functionNames = this.options.functionNames || DEFAULT_FUNCTIONS;
+	}
+
+	extractStringFromFunctionCall( args, functionName ) {
+		const strings = [];
+		switch ( functionName ) {
+			case '__':
+			case 'sprintf':
+			case '_n':
+				strings.push( args[ 0 ].value );
+				break;
+			case '_x':
+				strings.push( args[ 1 ].value + '\u0004' + args[ 0 ].value );
+				break;
+			case '_nx':
+				strings.push( args[ 3 ].value + '\u0004' + args[ 0 ].value );
+		}
+		return strings;
+	}
+
+	getStringsFromModule( module, extractor ) {
+		const source = module.originalSource().source();
+		if ( isEmpty( source ) ) {
+			return [];
+		}
+		let strings = [];
+		try {
+			const ast = recast.parse( source );
+			const { types } = recast;
+			types.visit( ast, {
```

Comment:
I'm following a common usage pattern found in many javascript libraries and non ee code I've worked with (using ESNext).  I'd love to see you review some of this libraries :)

With that said, if its a real pain for you I can modify.  But keep in mind your example is actually wrong.  If I dropped the use of the constant it would be `recast.types.visit`.

Regarding "mental gymnastics"... I know the feeling, that was the way it was for me when I was initially trying to grok some of these libraries.  But since I've gotten used to the syntax I actually find it a bit easier to work with.

One thing I probably _should_ do here instead of doing the extraction in this method is to do it at the beginning of the file right after the imports.

Reply:
OOPS! ya I had totally meant to write `recast.types.visit.` my bad!
Ya it might make it easier to grok if the extractions are done at the beginning and then you just have that set of variables to deal with in the rest of the file.
I hope you're right about it getting easier, because so far this JS stuff is so much more chaotic than PHP imho.